Add --bail option to handle errors thrown in generate command

ertrzyiks opened the pull request:
I've prepared an example repository when you can reproduce the bug: https://github.com/ertrzyiks/hexo-pug-syntax-error-example

Errors from the routing was not propagated in `hexo generate` command.
For example: the exit code was still 0 when the theme had syntax errors.

I use pug renderer. When I put 

```
link(rel="stylesheet"
```
The `hexo generate` should fail. It happily proceeds right now. This PR introduces `--bail` switch, which will make generate command to exit with an error in case of any error.

Example of usage:
```
hexo generate --bail
```

- [x] Add test cases for the changes.
- [ ] Passed the CI test.


<details>
<summary>The output of the current version</summary>

```
> yarn build
yarn build v0.27.5
$ hexo generate
INFO  Start processing
INFO  Files loaded in 67 ms
Unhandled rejection Error: /Users/mateuszderks/workspace/hexo-pug-syntax-error-example/themes/my-theme/layout/index.pug:2:1
    1| a(
  > 2| 
-------^

The end of the string reached with no closing bracket ) found.
    at makeError (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-error/index.js:32:13)
    at Lexer.error (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:58:15)
    at Lexer.bracketExpression (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:234:14)
    at Lexer.attrs (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1011:24)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.advance (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1356:15)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.getTokens (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1375:12)
    at lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:12:42)
    at Object.lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:99:27)
    at Function.loadString [as string] (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-load/index.js:44:24)
    at compileBody (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:86:18)
    at Object.exports.compile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:243:16)
    at pugCompile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:7:14)
    at pugRenderer (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:13:10)
    at Hexo.<anonymous> (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/index.js:12:10)

Unhandled rejection Error: /Users/mateuszderks/workspace/hexo-pug-syntax-error-example/themes/my-theme/layout/index.pug:2:1
    1| a(
  > 2| 
-------^

The end of the string reached with no closing bracket ) found.
    at makeError (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-error/index.js:32:13)
    at Lexer.error (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:58:15)
    at Lexer.bracketExpression (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:234:14)
    at Lexer.attrs (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1011:24)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.advance (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1356:15)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.getTokens (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1375:12)
    at lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:12:42)
    at Object.lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:99:27)
    at Function.loadString [as string] (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-load/index.js:44:24)
    at compileBody (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:86:18)
    at Object.exports.compile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:243:16)
    at pugCompile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:7:14)
    at pugRenderer (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:13:10)
    at Hexo.<anonymous> (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/index.js:12:10)

Unhandled rejection Error: /Users/mateuszderks/workspace/hexo-pug-syntax-error-example/themes/my-theme/layout/index.pug:2:1
    1| a(
  > 2| 
-------^

The end of the string reached with no closing bracket ) found.
    at makeError (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-error/index.js:32:13)
    at Lexer.error (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:58:15)
    at Lexer.bracketExpression (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:234:14)
    at Lexer.attrs (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1011:24)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.advance (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1356:15)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.getTokens (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1375:12)
    at lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:12:42)
    at Object.lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:99:27)
    at Function.loadString [as string] (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-load/index.js:44:24)
    at compileBody (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:86:18)
    at Object.exports.compile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:243:16)
    at pugCompile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:7:14)
    at pugRenderer (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:13:10)
    at Hexo.<anonymous> (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/index.js:12:10)

Unhandled rejection Error: /Users/mateuszderks/workspace/hexo-pug-syntax-error-example/themes/my-theme/layout/index.pug:2:1
    1| a(
  > 2| 
-------^

The end of the string reached with no closing bracket ) found.
    at makeError (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-error/index.js:32:13)
    at Lexer.error (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:58:15)
    at Lexer.bracketExpression (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:234:14)
    at Lexer.attrs (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1011:24)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.advance (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1356:15)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.getTokens (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1375:12)
    at lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:12:42)
    at Object.lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:99:27)
    at Function.loadString [as string] (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-load/index.js:44:24)
    at compileBody (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:86:18)
    at Object.exports.compile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:243:16)
    at pugCompile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:7:14)
    at pugRenderer (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:13:10)
    at Hexo.<anonymous> (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/index.js:12:10)

Unhandled rejection Error: /Users/mateuszderks/workspace/hexo-pug-syntax-error-example/themes/my-theme/layout/index.pug:2:1
    1| a(
  > 2| 
-------^

The end of the string reached with no closing bracket ) found.
    at makeError (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-error/index.js:32:13)
    at Lexer.error (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:58:15)
    at Lexer.bracketExpression (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:234:14)
    at Lexer.attrs (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1011:24)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.advance (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1356:15)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.getTokens (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1375:12)
    at lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:12:42)
    at Object.lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:99:27)
    at Function.loadString [as string] (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-load/index.js:44:24)
    at compileBody (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:86:18)
    at Object.exports.compile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:243:16)
    at pugCompile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:7:14)
    at pugRenderer (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:13:10)
    at Hexo.<anonymous> (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/index.js:12:10)

Done in 1.16s.

> echo $?
0
```
</details>

<details>
<summary>The output of the patched version</summary>

```
> yarn build
yarn build v0.27.5
$ hexo generate --bail
INFO  Start processing
INFO  Files loaded in 65 ms
FATAL Something's wrong. Maybe you can find the solution here: http://hexo.io/docs/troubleshooting.html
Error: /Users/mateuszderks/workspace/hexo-pug-syntax-error-example/themes/my-theme/layout/index.pug:2:1
    1| a(
  > 2| 
-------^

The end of the string reached with no closing bracket ) found.
    at makeError (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-error/index.js:32:13)
    at Lexer.error (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:58:15)
    at Lexer.bracketExpression (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:234:14)
    at Lexer.attrs (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1011:24)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.advance (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1356:15)
    at Lexer.callLexerFunction (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1319:23)
    at Lexer.getTokens (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:1375:12)
    at lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-lexer/index.js:12:42)
    at Object.lex (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:99:27)
    at Function.loadString [as string] (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug-load/index.js:44:24)
    at compileBody (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:86:18)
    at Object.exports.compile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/pug/lib/index.js:243:16)
    at pugCompile (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:7:14)
    at pugRenderer (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/lib/renderer.js:13:10)
    at Hexo.<anonymous> (/Users/mateuszderks/workspace/hexo-pug-syntax-error-example/node_modules/hexo-render-pug/index.js:12:10)
error Command failed with exit code 2.

> echo $?
1
```
</details>

ertrzyiks commented:
I will try to add tests for it, any guidance how to do that is welcome.

ertrzyiks commented:
@NoahDragon I updated the PR description, added output examples in the collapsible blocks at the bottom.

I noticed that this fix patches only some cases. When I use `hexo-renderer-pug`, it still does not report error correctly. This patch fixes usage with `hexo-render-pug`. I'm quite new to Hexo, so I'm not sure what to think about it yet. 

Probably, the follow-up PR fixing renderer case is needed.

NoahDragon commented:
Emmm, after reviewing the output sample, I feel this change may be good and bad. The good side is force user to change, instead of getting a blank page due to template rendering failure. The bad one is the hard stop may not necessary in some scenarios. E.g. an "undefined" error in a template file, even though it is an error, but it doesn't affect the site since the template only applies on certain pages.

I have a concern that this change may break many CI builds, not all of them are error-free (at least mine isn't :( ). I think I gonna to leave this PR open for now, and come back later when needed.

Thank you for the awesome work. Please feel free to discuss if my concern doesn't make sense.

ertrzyiks commented:
@NoahDragon Thanks for reviewing this PR. I see your concern. Actually, it is highly possible that people build their projects using silent failure as a feature and having a hard blocker now is not a nice UX at all.

I noticed this issue on my CI when I made a syntax error in my theme, the build was green and half of my blog was not rendered at all. We have a root cause fixed here, but what I really need is an option to check the quality of the updates. Maybe it could land as hidden behind a flag, like this?

```
hexo generate --bail
```

Of I could achieve the same effect by writing a plugin to bail? I will try to secure my blog somehow without breaking changes. Anyways, feel free to ping me back if you consider flag option as a mergeable solution.

NoahDragon commented:
@ertrzyiks You are right, maybe we could make this as a setting or parameter. That will be safe for existing users. 👍

JLHwung commented:
@ertrzyiks The `--bail` option looks good to me. Could you draft a PR working on adding `--bail` options? Thank you.

ertrzyiks commented:
@JLHwung sure, with pleasure. I have some free time this week.

ertrzyiks commented on lib/hexo/router.js:
```diff
@@ -155,6 +155,7 @@ RouteStream.prototype._read = function() {
     }
   }).catch(function(err) {
     self.emit('error', err);
+    self.push(null);
```
Without this line, tests hang forever in the scenario with failing renderer.

JLHwung replied:
IMO, this line is speculative, we should investigate the culprit of test hang.

ertrzyiks replied:
Without pushing null the stream never ends, that's the root cause of test hang. I will check how it behaves outside of test - maybe scripts is terminated in a different way, like process.exit()

ertrzyiks replied:
With this `push(null)` we can see the summary from the command
> INFO  0 files generated in 34 ms

Without it, the command just stops execution with unhandled rejection error printed to console.

In all other cases we close the stream, except this single case of `catch`. I think this line should be here since the beginning and it's actually a bug fix. 
What do you think @JLHwung ?

JLHwung replied:
@ertrzyiks I agree that the `stream` should be ended on any error. That is really keen observation. 👍

ertrzyiks commented on lib/plugins/console/generate.js:
```diff
@@ -154,13 +157,25 @@ function generateConsole(args) {
+
+  if (ignoreErrors) {
+    var noop = () => {}
+    dataStream.on('error', noop)
```
This is for tests because Stream seems to detect missing lack of `error` handler. I need to check how it affects functionality.

ertrzyiks replied:
Ok, replaced noop with a `log.error` to keep the current behavior.

ertrzyiks commented on lib/plugins/console/generate.js:
```diff
@@ -154,13 +157,25 @@ function generateConsole(args) {
+  if (ignoreErrors) {
+    var noop = () => {}
+    dataStream.on('error', noop)
+    return dataStream.pipe(new PassThrough())
```
`PassThrough` will pass data ignoring errors and this newly added `.on('error', reject)` will never be called in this case.

ertrzyiks commented:
@NoahDragon I rebased this branch and squashed commits, I'm not sure what to do about `codeclimate` check that failed though. Should I try to address it in my PR?

NoahDragon commented:
@ertrzyiks I think we could ignore the errors report from codeclimate. Could you please fix the eslint issue report in travis-ci?

ertrzyiks commented:
@NoahDragon done, I also updated the PR description

JLHwung commented on lib/plugins/console/generate.js:
```diff
@@ -44,11 +46,12 @@ function generateConsole(args) {
   function writeFile(path, force) {
     var dest = join(publicDir, path);
     var cacheId = 'public/' + path;
+    var dataStream = wrapDataStream(route.get(path), {ignoreErrors: !bail});
```
consider simplify the `wrapDataStream` to 
```
wrapDataStream(path, { bail })
```
so we don't have to remember the relationship between `ignoreErrors` and `bail`.

ertrzyiks replied:
makes sense 👍

JLHwung commented on test/scripts/console/generate.js:
```diff
@@ -168,4 +168,67 @@ describe('generate', () => {
   }));
+
+  it('proceeds after error when bail option is not set', () => {
+    hexo.extend.renderer.register('err', 'html', () => Promise.reject(new Error('Testing unhandled exception')));
```
Will the error populate in this situation?
```
hexo.extend.renderer.register('syncError', 'html', () => { 
throw new Error('Testing unhandled exception')
});
```

ertrzyiks replied:
Specs are green when I replace rejected promise with `throw new Error`
```
hexo.extend.renderer.register('err', 'html', () => {
      throw new Error('Testing unhandled exception')
    });
```

I'm not sure what did you mean by `syncError`, with the following code
```
hexo.extend.renderer.register('syncError'
```
that callback is never called.
